---
title: "DNA methylation workflow - workshop 2023 part2"
author: "Frédéric Silvestre"
date: "13/1/2023"
output: html_document
---

This part aims to learn the workflow of differential DNA methylation analysis using Bioconductor packages, such as methylKit. We will work on example data provided by the package.


# 0° Preparation of the working environment

## General setup of RMarkdown chunks and common shortcuts (for MacOS)

```{r setup}
if(!requireNamespace("knitr", quietly=TRUE))
        install.packages("knitr")
library(knitr)
if(!requireNamespace("rmarkdown", quietly=TRUE))
        install.packages("rmarkdown")
library(rmarkdown)

knitr::opts_chunk$set(eval=F, results="hide", echo = TRUE, include=TRUE, warning=TRUE, error=TRUE)
```

Change the general options in each chunk: eval=T (run the chunk); results=markup (show the results) or hide ; echo=T (show the code) ; include=T (include the chunk)    
Option/Command I to create a chunk  
Shift/option ( to [  
Option ( to {  
| shift/option L  
~ option n  
double space to change the paragraph  
to open /var folder > command + shift + G (write /var)  

## Set working directory 

```{r eval=T}
getwd()
setwd("***") # adapt to your computer
ls()
rm(list = ls())
tempdir() #where to save all the data from R
dev.off() #to reset the graph device options
sessionInfo()

```

## Install R v4.2 and update RStudio


## Install Bioconductor 3.16 and requested packages

Bioconductor: https://www.bioconductor.org/
search for available packages: https://www.bioconductor.org/packages/release/BiocViews.html#___Software

```{r}
# Check the version of BiocManager on your computer
BiocManager::version()

if(!requireNamespace("BiocManager", quietly=TRUE))
        install.packages("BiocManager")
BiocManager::install(version="3.16")

library(BiocManager)

#To check the number of packages available in Bioconductor
avail <- BiocManager::available()
length(avail)

#update your packages
BiocManager::install()

install.packages("stringi",type="mac.binary") #if an error message appears (aucun package nommé ‘stringi’ n'est trouvé)
library(stringi)

if(!requireNamespace("pacman", quietly=TRUE)) #the package pacman allows to install and load several packages from Cran at once if they're not installed.
        install.packages("pacman")
library(pacman)

BiocManager::install(c("methylKit", "ShortRead", "GenomicRanges", "genomation"))
pacman::p_load(methylKit, ShortRead, GenomicRanges,genomation, remotes, factoextra)

BiocManager::install("methylKit", force=T) # if it does not work

```

## To obtain help

```{r}

?mean
help.start()
help(package = "methylKit")
vignette("methylKit")
browseVignettes("methylKit")

```

## To save and load R objects

```{r}

saveRDS(tiles, file="tiles1.RDS")
b <- readRDS("tiles1.RDS")

```


# 1° Working on DNA methylation data with methylKit package

## Data filtering and exploratory analyses

We start the analysis with the methylation call file after the alignment with Bismark.
Create a file.list object from the txt files. System.file means the file where the packages are installed: This function is meant to intercept calls to system.file, so that it behaves well with packages loaded by devtools. It is made available when a package is loaded with load_all.
Example here with the package methylKit. Two files for the control and 2 for the treatment.

The location is: [1] "/Library/Frameworks/R.framework/Versions/4.2/Resources/library/methylKit/extdata/test1.myCpG.txt"
Press option while clicking on go to in the finder (Bibliothèque = Library). It will show the hidden files.
Or press fn + Command + shift + . in the finder to show the hidden files.

```{r}

system.file()#to find your system file

a <- system.file("extdata", "test1.myCpG.txt", package="methylKit")
a

file.list <- list( system.file("extdata", 
                            "test1.myCpG.txt", package = "methylKit"),
                system.file("extdata",
                            "test2.myCpG.txt", package = "methylKit"),
                system.file("extdata", 
                            "control1.myCpG.txt", package = "methylKit"),
                system.file("extdata", 
                            "control2.myCpG.txt", package = "methylKit") )

file.list
```

methRead function.
Now we can read the files and transform it into a **methylRawList** object (=list of methylRaw objects).
Tab-separated bedgraph like formats from Bismark methylation caller can also be read in by methylkit. In those cases, we have to provide either pipeline="bismarkCoverage" or pipeline="bismarkCytosineReport" to methRead function.
According to Liu 2020, a 30X coverage is the ideal balance.

```{r}
myobj <- methRead(file.list,
           sample.id=list("test1","test2","ctrl1","ctrl2"),
           assembly="hg18",
           treatment=c(1,1,0,0),
           context="CpG",
           dbdir = getwd(),
           mincov = 30
           )

#characterize myobj
length(myobj)
class(myobj)
head(myobj)
myobj
?methRead

#change the parameters
# how many rows are there using mincov of 10 ?

#****

```

Wa can change the arguments: header, context, resolution, dbdir
Further quality check from the read files.
Distribution of the %methylation.

```{r}

getMethylationStats(myobj[[1]],plot=T,both.strands=F)

?getMethylationStats

```

We plot the coverage values (we usually keep fragments with cov > 10 but a too high value means problems with the PCR).

```{r}

getCoverageStats(myobj[[1]],plot=T,both.strands=FALSE)

#what is the percentage of bases covered more than 526 times for sample 1 ?

#***
```

We should now discard the fragments with lower than 10 coverage and too high coverage (> 99.9% percentile of the sample).

```{r}
filtered.myobj <- filterByCoverage(myobj,lo.count=10,lo.perc=NULL,
                                      hi.count=NULL,hi.perc=99.9)
?filterByCoverage

# What is now the 99.9% percentile coverage ?

#***
```


Merging the files in a single object.

Setting destrand=TRUE (the default is FALSE) will merge reads on both strands of a CpG dinucleotide. This provides better coverage, but only advised when looking at CpG methylation (for CpH methylation this will cause wrong results in subsequent analyses). In addition, setting destrand=TRUE will only work when operating on base-pair resolution, otherwise setting this option TRUE will have no effect.

```{r}

meth <- methylKit::unite(filtered.myobj, destrand=T)
meth
?methylKit::unite

```


By default, unite function produces bases/regions covered in all samples. That requirement can be relaxed using “min.per.group” option in unite function.

```{r}
# creates a methylBase object, 
# where only CpGs covered with at least 1 sample per group will be returned

# there were two groups defined by the treatment vector, 
# given during the creation of myobj: treatment=c(1,1,0,0)

#****

# what is the location of the first CpG of the table ?

```

We can analyse the coefficient of variability for each group.

```{r}

pm=percMethylation(meth) # get percent methylation matrix ; only on CpG in all samples
head(pm, 10)

#What are the average CpG %methylation in CTL1 sample ?

#***

#calculate coefficient of variation for each CpG in CTL and in test

CV <- function(x){sd(x)/mean(x)*100}

CTL <- apply(pm[,3:4], MARGIN=1,FUN = CV)
summary(CTL)
head(CTL)
CTL

?apply

test <- apply(pm[,1:2], MARGIN=1,FUN = CV)
summary(test)

# What is the average CV for all groups together ?

#****

hist(CTL)

```

Filtering CpG
To remove different CpG based on criteria. For example, we can keep only the CpG with a minimum variation (and not the ones showing mutation C > T).
Below, we extract percent methylation values from CpGs as a matrix. Calculate standard deviation for each CpG and filter based on standard deviation. We also plot the distribution of per CpG standard deviations.

```{r}

pm=percMethylation(meth) # get percent methylation matrix
head(pm)

mds=matrixStats::rowSds(pm) # calculate standard deviation of CpGs
summary(mds)

hist(mds,col="cornflowerblue",xlab="Std. dev. per CpG")

#subsample CpG with a SD higher than 10 and compare the histogram of SD distribution with non-filtered CpG

#****

pm10 <- percMethylation(methsd10)
mds10=matrixStats::rowSds(pm10) # calculate standard deviation of CpGs

par(mfrow=c(1,2))
hist(mds10,col="red",xlab="Std. dev. per CpG", xlim=c(0,50))
hist(mds,col="cornflowerblue",xlab="Std. dev. per CpG")
par(mfrow=c(1,1))

```


Now, let’s assume we know the locations of C->T mutations. These locations should be removed from the analysis as they do not represent bisulfite treatment associated conversions. Mutation locations are stored in a GRanges object, and we can use that to remove CpGs overlapping with mutations. In order to do overlap operation, we will convert the methylKit object to a GRanges object and do the filtering with %over% function within [ ]. The returned object will still be a methylKit object.

```{r}

mut=GRanges(seqnames=c("chr21","chr21"),
            ranges=IRanges(start=c(9860126, 9944505),
                           end=c(9860126, 9944505)))

# select CpGs that do not overlap with mutations
sub.meth=meth[! as(meth,"GRanges") %over% mut,]
nrow(meth)
nrow(sub.meth)
```


## PCA and clustering 

Only on a matrix with no NA (CpG in all samples)

Search for correlations between samples:

```{r}

getCorrelation(meth,plot=TRUE)
getCorrelation(meth,plot=F)

```

Clustering samples to group data points by their similarity.

```{r}

clusterSamples(meth, dist="correlation", method="ward", plot=TRUE)

?clusterSamples

#Setting the plot=FALSE will return a dendrogram object which can be manipulated by users or fed in to other user functions that can work with dendrograms.

hc = clusterSamples(meth, dist="correlation", method="ward.D", plot=FALSE)
hc

```


Perform PCA.
Here we plot a screen plot for importance of components.

```{r}

pc <- PCASamples(meth,obj.return = T, adj.lim=c(1,1), comp = c(1,2))
pc

PCASamples(meth, obj.return=F)

fviz_pca_biplot(pc, repel = TRUE)

# Eigenvalues
eig.val <- get_eigenvalue(pc)
eig.val

fviz_eig(pc)

?PCASamples
```


In this case, we also returned an object from the plotting function. This is the output of R prcomp function, which includes loadings and eigen vectors which might be useful. You can also do your own PCA analysis using percMethylation and prcomp. In the case above, the methylation matrix is transposed. This allows us to compare distances between samples on the PCA scatterplot.


## Batch control

We have implemented some rudimentary functionality for batch effect control. You can check which one of the principal components are statistically associated with the potential batch effects such as batch processing dates, age of subjects, sex of subjects using assocComp. The function gets principal components from the percent methylation matrix derived from the input methylBase object, and checks for association. The tests for association are either via Kruskal-Wallis test or Wilcoxon test for categorical attributes and correlation test for numerical attributes for samples such as age. If you are convinced that some principal components are accounting for batch effects, you can remove those principal components from your data using removeComp.

```{r}
# make some batch data frame
# this is a bogus data frame
# we don't have batch information
# for the example data

sampleAnnotation=data.frame(batch_id=c("a","a","b","b"),
                            age=c(19,37,35,17))

sampleAnnotation

as=assocComp(mBase=meth,sampleAnnotation)
as

?assocComp

#if we want to remove one component:
newObj=removeComp(meth,comp=3)
?removeComp
newObj
as2=assocComp(mBase=newObj,sampleAnnotation)
as2

#we can't do the analysis if we have NA > use the meth before filtering
is.na(meth)

```



## Analysis on regions

For some situations, it might be desirable to summarize methylation information over tiling windows rather than doing base-pair resolution analysis. methylKit provides functionality to do such analysis. The function below tiles the genome with windows of 1000bp length and 1000bp step-size and summarizes the methylation information on those tiles. In this case, it returns a methylRawList object which can be fed into unite and calculateDiffMeth functions consecutively to get differentially methylated regions. The tilling function adds up C and T counts from each covered cytosine and returns a total C and T count for each tile.

As mentioned before, methRead sets a minimum coverage threshold of 10 reads per cytosine to ensure good quality for downstream base-pair resolution analysis. However in the case of tiling window / regional analysis one might want to set the initial per base coverage threshold to a lower value and then filter based on the number of bases (cytosines) per region. Filtering samples based on read coverage might still be appropriate to remove coverage biases.

```{r}

myobj_lowCov = methRead(file.list,
           sample.id=list("test1","test2","ctrl1","ctrl2"),
           assembly="hg18",
           treatment=c(1,1,0,0),
           context="CpG",
           mincov = 3
           )

tiles = tileMethylCounts(myobj,win.size=1000,step.size=1000,cov.bases = 0)
?tileMethylCounts

#How many CpG are there on tiles number 4 ?

#***

# unite the different samples

tile <- methylKit::unite(tiles, destrand=FALSE, min.per.group = NULL)

tile
head(tiles)

```


## 2° Extracting regions of interest

Search for differentially methylated CpG sites or regions.
Usually using Fisher's exact test and methylKit package or logistic regression. Here we show for one sample per group (Fisher).
calculateDiffMeth: The function calculates differential methylation statistics between two groups of samples. The function uses either logistic regression test or Fisher's Exact test to calculate differential methylation.
We can add overdispersion, covariate and the method of adjustment (SLIM by deffault, or BH, bonferroni, fdr, qvalue).

```{r}

getSampleID(meth)
new.meth <- reorganize(meth,sample.ids=c("test1","ctrl1"),treatment=c(1,0))
new.meth

dmf <- calculateDiffMeth(new.meth, overdispersion = NULL, covariate=NULL, adjust = "SLIM", test="fast.fisher")
dim(dmf)
dmf

?calculateDiffMeth
```

As mentioned, we can also pool the samples from the same group by adding up the number of Cs and Ts per group. This way even if we have replicated experiments we treat them as single experiments, and can apply Fisher’s exact test. We will now pool the samples and apply:


```{r}
pooled.meth=pool(meth,sample.ids=c("test","control"))
dm.pooledf=calculateDiffMeth(pooled.meth)
dm.pooledf
```

CalculateDiffMeth function returns the P-values for all bases or regions in the input methylBase object. We need to filter to get differentially methylated CpGs. This can be done via getMethlyDiff function or simple filtering via [ ] notation. Below we show how to filter the methylDiff object output by calculateDiffMeth() function in order to get differentially methylated CpGs. The function arguments defines cutoff values for the methylation difference between groups and Q-value. In these cases, we require a methylation difference of 25% and Q-value of at least 0.01.

```{r}
?getMethylDiff

# get differentially methylated bases/regions with specific cutoffs

all.diff=getMethylDiff(dmf,difference=10,qvalue=0.01,type="all")
all.diff

# get hyper-methylated
hyper=getMethylDiff(dmf,difference=25,qvalue=0.01,type="hyper")
hyper

# get hypo-methylated
hypo=getMethylDiff(dmf,difference=25,qvalue=0.01,type="hypo")
hypo

#using [ ] notation
hyper2=dm.pooledf[dm.pooledf$qvalue < 0.01 & dm.pooledf$meth.diff > 25,]
hyper2
```

Logistic regression based tests.
However, linear-regression-based methods might produce fitted methylation levels outside the range  
[0,1] unless the values are transformed before regression. An alternative is logistic regression, which can deal with data strictly bounded between 0 and 1 and with non-constant variance, such as methylation proportion/fraction values.
If the observed variance is larger or smaller than assumed by the model, one speaks of under- or over-dispersion. This over/under-dispersion can be corrected by calculating a scaling factor and using that factor to adjust the variance estimates.
MethylKit can apply logistic regression to test the methylation difference with or without the over-dispersion correction. In this case, Chi-square or F-test can be used to compare the difference in the deviances of the null model and the alternative model. The null model assumes there is no relationship between sample groups and methylation, and the alternative model assumes that there is a relationship where sample groups are predictive of methylation values for a given CpG or region for which the model is constructed. Next, we are going to use the logistic regression based model with over-dispersion correction and Chi-square test. Cfr book.

Overdispersion occurs when the observed variance is higher than the variance of a theoretical model. 


*Exercise to change the parameters.*

```{r}

dm.lr=calculateDiffMeth(meth,overdispersion = "MN" ,test ="Chisq", covariate=NULL, adjust = "SLIM")
dm.lr

```

CalculateDiffMeth function returns the P-values for all bases or regions in the input methylBase object. We need to filter to get differentially methylated CpGs. This can be done via getMethlyDiff function or simple filtering via [ ] notation. Below we show how to filter the methylDiff object output by calculateDiffMeth() function in order to get differentially methylated CpGs. The function arguments defines cutoff values for the methylation difference between groups and Q-value. In these cases, we require a methylation difference of 25% and Q-value of at least 0.01.


```{r}

# get differentially methylated bases/regions with a cutoff of 20 and a qvalue of 0.05
# how many CpG are significant ?
# how many are hypo-methylated ?

#***

# get hypo-methylated

#***

```

More complex regression models use beta binomial distribution and are particularly useful for better modeling the variance. Similar to logistic regression, their observation follows binomial distribution (number of reads), but methylation proportion itself can vary across samples, according to a beta distribution. It can deal with fitting values in the (0,1) range and performs better when there is greater variance than expected by the simple logistic model. In essence, these models have a different way of calculating a scaling factor when there is over-dispersion in the model. Further enhancements are made to these models by using the empirical Bayes methods that can better estimate hyper parameters of the beta distribution (variance-related parameters) by borrowing information between loci or regions within the genome to aid with inference about each individual loci or region. 
We are now going to use a beta-binomial based model called DSS(Feng, Conneely, and Wu 2014) to calculate differential methylation.
Beta-binomial model models overdispersion but is more intensive. Can't use covariate (except with DSS.general).
Logistic is hard to incorporate overdispersion.

```{r}

dm.dss=calculateDiffMethDSS(meth, adjust = "SLIM")
dm.dss

# how many CpG are differentially methylated ?

#***

?calculateDiffMethDSS
```

We can perform the analysis by region.
The function below tiles the genome with windows 1000bp length and 1000bp step-size and summarizes the methylation information on those tiles. In this case, it returns a methylRawList object which can be fed into unite and calculateDiffMeth functions consecutively to get differentially methylated regions.

```{r}

tiles.dss=calculateDiffMethDSS(tile, adjust = "SLIM")
tiles.dss

tiles.diff=getMethylDiff(tiles.dss,difference=20,qvalue=0.05,type="all")
tiles.diff

```

In addition, if we are interested in particular regions, we can also get those regions as methylKit objects after summarizing the methylation information as described above. The code below summarizes the methylation information over a given set of promoter regions and outputs a methylRaw or methylRawList object depending on the input. We are using the output of genomation functions used above to provide the locations of promoters. For regional summary functions, we need to provide regions of interest as GRanges object.

```{r}

# read the gene BED file
gene.obj=readTranscriptFeatures(system.file("extdata", "refseq.hg18.bed.txt", 
                                           package = "methylKit"))
promoters=regionCounts(myobj,gene.obj$promoters)
?regionCounts
head(promoters[[1]])
gene.obj
```

In addition, it is possible to cluster DMCs based on their proximity and direction of differential methylation. This can be achieved by methSeg function in methylKit. We will see more about methSeg function in the following section. But it can take the output of getMethylDiff function therefore can work on DMCs to get differentially methylated regions.


## Adding covariates.

Covariates can be included in the analysis as well in methylKit. The calculateDiffMeth function will then try to separate the influence of the covariates from the treatment effect via the logistic regression model. In this case, we will test if full model (model with treatment and covariates) is better than the model with the covariates only. If there is no effect due to the treatment (sample groups), the full model will not explain the data better than the model with covariates only. In calculateDiffMeth, this is achieved by supplying the covariates argument in the format of a data.frame. Below, we simulate methylation data and add make a data.frame for the age. The data frame can include more columns, and those columns can also be factor variables. The row order of the data.frame should match the order of samples in the methylBase object. Below we are showing an example of this using a simulated data set where methylation values of CpGs will be affected by the age of the sample.

```{r}

covariates=data.frame(age=c(30,80,34,30,80,40))
sim.methylBase=dataSim(replicates=6,sites=1000,
                        treatment=c(rep(1,3),rep(0,3)),
                        covariates=covariates,
                        sample.ids=c(paste0("test",1:3),paste0("ctrl",1:3)))
?dataSim

my.diffMeth3=calculateDiffMeth(sim.methylBase,
                               covariates=covariates,
                               overdispersion="MN",
                               test="Chisq")
my.diffMeth3
getMethylDiff(my.diffMeth3,difference=10,qvalue=0.01,type="all")


covariates <- data.frame(age=c(30, 80, 30, 80))
my.diffMeth4=calculateDiffMeth(meth,
                               covariates=covariates,
                               overdispersion="MN",
                               test="Chisq",
                               adjust="SLIM")
my.diffMeth4
all.diff4=getMethylDiff(my.diffMeth4,difference=10,qvalue=0.01,type="all")
all.diff4

```


# 3° Annotation of DMR/DMC

Below, we are reading a BED file for transcripts and using that to annotate DMCs with promoter/intron/exon/intergenic annotation.genomation::readTranscriptFeatures function reads a BED12 file, calculates the coordinates of promoters, exons and introns and the subsequent function uses that information for annotation.

One of larger hindrances in computational genomics stems from the myriad of formats that are used to store the data. Although some formats have been selected as de facto standards for specific kind of biological data (e.g. BAM, VCF), almost all publications come with supplementary tables that do not have the same structure, but hold similar information. The tables usually have a tabular format, contain the location of elements in genomic coordinates and various metadata colums. genomation contais functions to read genomic intervals and genomic annotation provided they are in a tabular format. These functions will read the data from flat files into GRanges or GRangesList objects.

readGeneric is the workhorse of the genomation package. It is a function developed specifically for input of genomic data in tabular formats, and their conversion to a GRanges object. By default, the function persumes that the file is a standard .bed file containing columns chr, start, end.

```{r}

browseVignettes("genomation")

# read the gene BED file
transcriptBED=system.file("extdata", "refseq.hg18.bed.txt", 
                                           package = "methylKit")

readGeneric(transcriptBED, header=T, keep.all.metadata=F)

gene.obj=readTranscriptFeatures(transcriptBED)
class(gene.obj)
gene.obj

?readTranscriptFeatures

# annotate differentially methylated CpGs with 
# promoter/exon/intron using annotation data
#
annotateWithGeneParts(as(all.diff,"GRanges"),gene.obj)

?annotateWithGeneParts

```

Similarly, we can read the CpG island annotation and annotate our differentially methylated bases/regions with them.

````{r}

# read the shores and flanking regions and name the flanks as shores 
# and CpG islands as CpGi
cpg.file=system.file("extdata", "cpgi.hg18.bed.txt", 
                                        package = "methylKit")
cpg.obj=readFeatureFlank(cpg.file,
                           feature.flank.name=c("CpGi","shores"))
head(cpg.obj)
#
# convert methylDiff object to GRanges and annotate
diffCpGann=annotateWithFeatureFlank(as(all.diff,"GRanges"),
                                    cpg.obj$CpGi,cpg.obj$shores,
                         feature.name="CpGi",flank.name="shores")
diffCpGann
```

We can also summarize methylation information over a set of defined regions such as promoters or CpG islands. The function below summarizes the methylation information over a given set of promoter regions and outputs a methylRaw or methylRawList object depending on the input. We are using the output of genomation functions used above to provide the locations of promoters. For regional summary functions, we need to provide regions of interest as GRanges object.

```{r}
promoters=regionCounts(myobj,gene.obj$promoters)
head(promoters[[1]])
```

After getting the annotation of differentially methylated regions, we can get the distance to TSS and nearest gene name using the getAssociationWithTSS function from genomation package.

```{r}
diffAnn=annotateWithGeneParts(as(all.diff,"GRanges"),gene.obj)

# target.row is the row number in all.diff
a <- getAssociationWithTSS(diffAnn)
a

?getAssociationWithTSS
```

It is also desirable to get percentage/number of differentially methylated regions that overlap with intron/exon/promoters

```{r}

getTargetAnnotationStats(diffAnn,percentage=TRUE,precedence=TRUE)

```

We can also plot the percentage of differentially methylated bases overlapping with exon/intron/promoters

```{r}

plotTargetAnnotation(diffAnn,precedence=TRUE,
    main="differential methylation annotation")
```

We can also plot the CpG island annotation the same way. The plot below shows what percentage of differentially methylated bases are on CpG islands, CpG island shores and other regions.


```{r}

plotTargetAnnotation(diffCpGann,col=c("green","gray","white"),
       main="differential methylation annotation")
```

It might be also useful to get percentage of intron/exon/promoters that overlap with differentially methylated bases.

```{r}

getFeatsWithTargetsStats(diffAnn,percentage=TRUE)

```


We can also read gtf format files

```{r}

gff.file = system.file("extdata/chr21.refseq.hg19.gtf", package = "genomation")

gff = gffToGRanges(gff.file)

head(gff)
class(gff)

gffsplit=as(split(gff, gff$type), "GRangesList")
gffsplit[[1]]

all.diffG <- as(all.diff, "GRanges")

annotateWithFeatures(all.diffG,gffsplit)

annotateWithGeneParts(all.diffG,gffsplit)

?annotateWithGeneParts

```


