---
title: "Working with bioconductor - workshop 2023 part1"
author: "Frédéric Silvestre"
date: "11/1/2023"
output:
  html_document: default
  pdf_document: default
---

This first part of the workshop aims to introduce bioconductor packages and to work with sequences and genomic objects in R. It will also introduce Rmarkdown.


# 0° Preparation of the working environment

## General setup of RMarkdown chunks and common shortcuts (for MacOS)

```{r setup}
if(!requireNamespace("knitr", quietly=TRUE))
        install.packages("knitr")
library(knitr)
if(!requireNamespace("rmarkdown", quietly=TRUE))
        install.packages("rmarkdown")
library(rmarkdown)

knitr::opts_chunk$set(eval=F, results="hide", echo = TRUE, include=TRUE, warning=TRUE, error=TRUE)
```

Change the general options in each chunk: eval=T (run the chunk); results=markup (show the results) or hide ; echo=T (show the code) ; include=T (include the chunk)    
Option/Command I to create a chunk  
Shift/option ( to [  
Option ( to {  
| shift/option L  
~ option n  
double space to change the paragraph  
to open /var folder > command + shift + G (write /var)  

## Set working directory 

```{r eval=T}
getwd()
setwd("/Users/fsilvestre/bioinformatic_workshop/") # adapt to your computer
ls()
rm(list = ls())
tempdir() #where to save all the data from R
dev.off() #to reset the graph device options
sessionInfo()

```

## Install R v4.2 and update RStudio


## Install Bioconductor 3.16 and requested packages

Bioconductor: https://www.bioconductor.org/
search for available packages: https://www.bioconductor.org/packages/release/BiocViews.html#___Software

```{r}
# Check the version of BiocManager on your computer
BiocManager::version()

if(!requireNamespace("BiocManager", quietly=TRUE))
        install.packages("BiocManager")

BiocManager::install(version="3.16")

library(BiocManager)

#To check the number of packages available in Bioconductor
avail <- BiocManager::available()
length(avail)

#update your packages
BiocManager::install()

install.packages("stringi",type="mac.binary") #if an error message appears (aucun package nommé ‘stringi’ n'est trouvé)
library(stringi)

if(!requireNamespace("pacman", quietly=TRUE)) #the package pacman allows to install and load several packages from Cran at once if they're not installed.
        install.packages("pacman")
library(pacman)

BiocManager::install(c("GenomicRanges", "GenomeInfoDb", "IRanges", "Biostrings", "BSgenome", "AnnotationHub", "SummarizedExperiment", "airway", "GenomicFeatures", "SGSeq", "qvalue"))
pacman::p_load(GenomicRanges, GenomeInfoDb, IRanges, Biostrings, BSgenome, AnnotationHub, SummarizedExperiment, airway, GenomicFeatures, SGSeq, qvalue)

```

## To obtain help

```{r}

?mean
help.start()
help(package = "GenomicRanges")
vignette("GenomicRanges")
browseVignettes("GenomicRanges")

```

## To save and load R objects

```{r}

saveRDS(tiles, file="tiles1.RDS")
b <- readRDS("tiles1.RDS")

```



# 1° Working with sequences

## Create a DNAstring object

```{r}

DNA <- DNAString("TCAG") #create a Biostring object
DNA
class(DNA)

a <- as.character(DNA) #give the sequence
class(a)
a

length(DNA) #count the number of nucleotides
b <- DNA[2:3]
class(b)
b

DNA2 <- DNAString("AAGCTAT")
DNA2
```

## Create a set of DNAstring objects

```{r}

DNA3 <- DNAStringSet(c("TCA", "AAATCG")) 
DNA3
class(DNA3)
length(DNA3) #count the number of objects Biostring
width(DNA3) #count the number of nucleotides

set1 <- DNA3[1]
set1
width(set1)

unique(DNA3)
duplicated(DNA3)

DNA4 <- DNAStringSet(c("TCA", "AAATCG", "TCA")) 
unique(DNA4)
dup <- duplicated(DNA4)
sum(dup, na.rm=T) #count the number of duplicate sequences

```

## Combine 2 DNAString objects

```{r}

seqs1 <- DNAStringSet(list(seq1=DNA, seq2=DNA2)) #combine and give a name
seqs1
class(seqs1)
seqs1$seq1

```


## Seq in FASTA format of the gene CDR2 of Kmar (downloaded from https://www.ensembl.org/)

```{r}

?readDNAStringSet

seqs2 <- readDNAStringSet(filepath="RAW_data_Florida_Belize/Sequences/kmarCDR2.fa", format="fasta") 

seqs2
names(seqs2)
class(seqs2)
length(seqs2)
width(seqs2)

vmatchPattern("AGGAGGT", seqs2) #retrieve the motif in the sequence

subseq1 <- subseq(seqs2, 2205,2211)
subseq1
subseq1 <- subseq(seqs2, start=2205, end=2211)
subseq1
class(subseq1)

print(as.character(seqs2))

letterFrequency(seqs2, "T")
letterFrequency(seqs2,"GC") #G or C

#Calculate the GC contain
#****

dinucleotideFrequency(seqs2)

?dinucleotideFrequency

translate(seqs2)

#Subset the sequence between position 20 and 100
#****

```


## Combine 2 subsequences in a single DNAStringSet

```{r}

#combine the 2 sequences of CDR2 gene (seqs2 and subseq1) in 2 DNAStringSet object of length 1, and in 1 DNAStringSet object of length 2
#***
```



## Work on set of genes: GRanges

```{r}

?GRanges

myseq <- DNAString("ACCATTGATTAT")
myseq
myset <- DNAStringSet(list(chrI=myseq, chrII=DNAString("ACGTACGT")))
myset
class(myset)
range1 <- GRanges("chrI", IRanges(start = 3, end = 5), strand = "+")
range1
class(range1)

seqnames(range1)
strand(range1)
getSeq(myset, range1)

mcols(range1)$wobble <- 10 #create metadata columns
range1
seqinfo(range1)
range2 <- GRanges(c("chrZ","chrY"), IRanges(start=c(5,10), end=c(35,45)), strand="+")
range2
range3 <- c(range1, range2)
range3
range3[1:2,]

```


## Download gtf file from your species: http://www.ensembl.org/Kryptolebias_marmoratus/Info/Index

https://www.ensembl.org/info/website/upload/gff.html 
To find all the files to download here: https://www.ensembl.org/info/data/ftp/index.html

```{r}

features <- import("RAW_data_Florida_Belize/Sequences/GTF/genomic.gtf")
features
mcols(features) <- mcols(features)[,c("type","gene","gene_id")]
unique(features$gene)
features
class(features)
feat <- subset(features, gene=="mtx3")
feat

```


## To find references for different species using AnnotationHub

```{r}

package = "AnnotationHub"

#As of AnnotationHub (>2.23.2), default caching location has changed. Run the code below to adapt the cache
oldcache = path.expand(rappdirs::user_cache_dir(appname=package))
    setAnnotationHubOption("CACHE", oldcache)
    ah = AnnotationHub(localHub=TRUE)
    ## removes old location and all resources
    removeCache(ah, ask=FALSE)

    ## create the new default caching location
    newcache = tools::R_user_dir(package, which="cache")
    setAnnotationHubOption("CACHE", newcache)
    ah = AnnotationHub()

    
browseVignettes("AnnotationHub")

ah <- AnnotationHub()
length(ah)

colnames(mcols(ah))
unique(ah$species)

records <- query(ah,"Kryptolebias marmoratus")
records
table(records$genome)
table(records$sourcetype)
mcols(records)[,c("title", "rdataclass", "sourcetype")]
km <- ah[["AH64640"]]

```


## Working with expressionSet

```{r}

read.table("RAW_data_Florida_Belize/txt/expressiondata.txt", header=TRUE, sep="\t", row.names=1)
exprs <- as.matrix(read.table("RAW_data_Florida_Belize/txt/expressiondata.txt", header=TRUE, sep="\t", row.names=1))
exprs
class(exprs)
dim(exprs)
colnames(exprs)
head(exprs[,1:3])
head(exprs[1:3,])
head(exprs[1:3])

minimalSet <- ExpressionSet(assayData=exprs)
minimalSet
pData <- read.table("RAW_data_Florida_Belize/txt/pData.txt", row.names=1, header=TRUE, sep="\t")
pData
summary(pData)

all(rownames(pData)==colnames(exprs))

names(pData)
class(pData)
sapply(pData, class)
pData[c(2, 4), c("Sex")]
pData[pData$Age>20,]

phenoData <- new("AnnotatedDataFrame", data=pData)
phenoData

exampleSet <- ExpressionSet(assayData=exprs, phenoData=phenoData)
exampleSet
exampleSet$Sex[1:3]
featureNames(exampleSet)
sampleNames(exampleSet)
varLabels(exampleSet)

mat <- exprs(exampleSet)
mat
males <- exampleSet[,exampleSet$Sex=="M"]
males
exprs(males)

```


## SummarizedExperiment for RNAseq

```{r}

data(airway, package="airway")
se <- airway
se
assays(se)
assays(se)$counts
rowRanges(se)
colData(se)
se[,se$dex=="trt"]
metadata(se)
```

## Correction for multiple comparisons


```{r}

?qvalue # proportion of false positive when the test is positive ; compare to Bonferoni and FDR

data(hedenfalk)
qvalues <- qvalue(hedenfalk$p, fdr.level=0.05, method="bootstrap")$q
qvalues

bonferroni <- p.adjust(hedenfalk$p, method="bonferroni")
fdr <- p.adjust(hedenfalk$p, method="fdr")

plot(hedenfalk$p,qvalues,pch=19,ylim=c(0,1),
     xlab="raw P-values",ylab="adjusted P-values")
points(hedenfalk$p,bonferroni,pch=19,col="red")
points(hedenfalk$p,fdr,pch=19,col="blue")
legend("bottomright",legend=c("q-value","FDR (BH)","Bonferroni"),
       fill=c("black","blue","red"))
```

```{r}

sessionInfo()

```


